server/src/main/java/org/elasticsearch/index/engine/RecoverySourcePruneMergePolicy.java

```diff
@@ -212,38 +196,76 @@ public long ramBytesUsed() {
             }
         }
 
-        private static class FilterStoredFieldsReader extends StoredFieldsReader {
+        private abstract static class FilterStoredFieldsReader extends StoredFieldsReader {
 
-            private final StoredFieldsReader fieldsReader;
+            protected final StoredFieldsReader in;
 
             FilterStoredFieldsReader(StoredFieldsReader fieldsReader) {
-                this.fieldsReader = fieldsReader;
+                this.in = fieldsReader;
             }
 
             @Override
             public long ramBytesUsed() {
-                return fieldsReader.ramBytesUsed();
+                return in.ramBytesUsed();
             }
 
             @Override
             public void close() throws IOException {
-                fieldsReader.close();
+                in.close();
             }
 
             @Override
             public void visitDocument(int docID, StoredFieldVisitor visitor) throws IOException {
-                fieldsReader.visitDocument(docID, visitor);
+                in.visitDocument(docID, visitor);
             }
 
             @Override
-            public StoredFieldsReader clone() {
-                return fieldsReader.clone();
-            }
+            public abstract StoredFieldsReader clone();
```

Comment:
Maybe also make `getMergeInstance()` method abstract here? So that implementors have to think about whether that method should be implemented too?

Reply:
The consensus on Lucene is that FilterXXX classes should not override methods that have a default implementation in the base class (e.g. FilterLeafReader doesn't override `LeafReader#postings`), so I tried to be consistent.